Refactor contact module

schmengler opened the pull request:
This PR contains:

- removal of Zend Framework 1 dependencies
- usage of result factories in controller
- show meaningful error messages to user if possible
- separated `sendEmail` and `validatedParams` methods for easier rewriting

     I intended to move it out of the controller altogether but it seems to be against what the team had in mind with the abstract controller

Why?

- In Magento 1 the contact module was a popular one to take as an example for how a simple module should work. Let's keep it aligned with current best practices
- Having coding validation and email sending all in the `execute` method makes extending it hard (as stated above, I'd move it out of the controller but did not want to take the PR too far)

schmengler committed:
Remove Zend_Validate calls

schmengler committed:
Replace Zend_Validate_EmailAddress with 100% correct email validation

Reference: https://hackernoon.com/the-100-correct-way-to-validate-email-addresses-7c4818f24643

schmengler committed:
Extract sendEmail method in contact form controller

schmengler committed:
Extract validatedParams method in contact form controller, use redirect result factory

schmengler committed:
Show meaningful error message if contact form validation fails

schmengler committed:
Make validation and sending methods in contacts form controller protected

I intended to move it out of the controller altogether but it seems to be against what the team had in mind with the abstract controller

dmanners commented:
I think I love you. Death to Zend!

fooman commented on app/code/Magento/Contact/Controller/Index/Post.php:
```diff
@@ -101,4 +63,57 @@ private function getDataPersistor()
+        if (trim($request->getParam('comment')) === '') {
+            throw new LocalizedException(__('Comment is missing'));
+        }
+        if (false === \strpos($request->getParam('email'), '@')) {
```
You are losing quite a bit of functionality here vs Zend_Validate_EmailAddress. Maybe the better approach would be to switch to the ZF2 component https://github.com/zendframework/zend-validator/blob/master/src/EmailAddress.php#L492

schmengler replied:
I did that on purpose because I don't believe in email address validation via regular expressions (relevant: https://hackernoon.com/the-100-correct-way-to-validate-email-addresses-7c4818f24643). If there's a decision to use the MX check, the ZF2 email validator would be useful again.

fooman replied:
Great article - you got me convinced.

joe-vortex replied:
Why are we not using php's native `filter_var` function? Surely this would be better as:
```
if (!filter_var($request->getParam('email'), FILTER_VALIDATE_EMAIL)) {
    throw new LocalizedException(__('Invalid email address'));
}
```

schmengler replied:
@joe-vortex same reason that I dropped Zend_Validate_EmailAddress. The top comment on http://php.net/manual/en/function.filter-var.php already gives an example

schmengler committed:
Add docblock comments

schmengler committed:
Update unit test

schmengler committed:
Fix code style violations

PieterCappelle commented:
Great PR!

vrann commented:
@schmengler thank you for contribution, the perfect example of good pull request!
Let me come back on this if there will be any additional questions upon verification.

vrann commented on app/code/Magento/Contact/Controller/Index/Post.php:
```diff
@@ -101,4 +65,61 @@ private function getDataPersistor()
+            ->setTemplateOptions(
+                [
+                    'area' => \Magento\Backend\App\Area\FrontNameResolver::AREA_CODE,
+                    'store' => \Magento\Store\Model\Store::DEFAULT_STORE_ID,
```
our static test failing with this message:
```
Module Magento\Contact has undeclared dependencies: hard [Magento\Backend]
```
do we really need the dependency on Backend here? If yes, please add the dependency to the composer.json otherwise.

schmengler replied:
This is the same code as before, just moved to a new method, but good that the static tests catched it :) I'll check if we can get rid of the dependency!

vrann replied:
yeah, noticed that, though the fact you touched that classed triggered static test failure :) I'd appreciate if you ensure that the dependency is needed and add it to the composer.

vrann commented on app/code/Magento/Contact/Controller/Index/Post.php:
```diff
@@ -101,4 +65,61 @@ private function getDataPersistor()
+     * @param array $post Post data from contact form
+     * @return void
+     */
+    protected function sendEmail($post)
```
Magento discourages the introduction of any new "protected" properties or methods. The reason is that we are trying to get rid of inheritance-based APIS and promote composition ver inheritance. Only public or private modifiers should be used.

Please change this to "private"

schmengler replied:
That's good to hear. Would you mind if I removed the abstract base controller then and moved the email sending out of the controller? I hesitated because it looked like inheritance was *intended* here and extensions might already use it (although that's improbable)

vrann replied:
Abstract base controllers (and similar framework pieces such as Blocks, Models, etc) still remain places where legacy inheritance-based APIs haven't been refactored yet. I would suggest keeping abstract base controller for consistency with other controllers (and you are right, for backward compatibility)

However the policy for the new methods/properties to be public or private.

schmengler replied:
Just to be clear, I was referring to this one, not the framework base controller: https://github.com/magento/magento2/blob/develop/app/code/Magento/Contact/Controller/Index.php

vrann commented on app/code/Magento/Contact/Controller/Index/Post.php:
```diff
@@ -19,72 +22,33 @@ class Post extends \Magento\Contact\Controller\Index
-        if (!$post) {
-            $this->_redirect('*/*/');
-            return;
+        if (! $this->isPostRequest()) {
```
extra space here after "!"

vrann reviewed (changes requested):
Added few comments based on code review and static tests

vrann commented:
@schmengler let me know do you have time to work on requested changes. If you don't at the moment, I'll apply fixes and get it merged.

schmengler commented:
Thank you! I'll try to get it done this week and will let you know if it does not work out.

schmengler committed:
Revert "Make validation and sending methods in contacts form controller protected"

This reverts commit c8fbb7072715c3d59e62ad12e37406518f2c0b14.

schmengler committed:
Remove unnecessary dependency

while constants should be preferred over magic strings, the dependency to FrontNameResolver seems arbitrary, just because it happens to have a constant for 'adminhtml' defined.

schmengler committed:
Moved down unnecessary dependencies from abstract base controller in contacts module

Additionally, the tests have been refactored for current coding standards and PHPUnit forward compatibility

schmengler committed:
Move configuration access to config model / API interface

original constants and methods are still in place for backwards compatibility

schmengler committed:
Merge remote-tracking branch 'origin/develop' into contact-refactor

Conflicts:
	app/code/Magento/Contact/Test/Unit/Controller/Index/IndexTest.php

schmengler committed:
Extract mail sending from contacts controller, introduce mail service contract

schmengler commented:
I made the changes, and some more, decided to extract the mail sending functionality from the controller eventually.

I also introduced two API interfaces:

- `\Magento\Contact\Api\ConfigInterface` allows reuse of the contact form with different email template, sender and recipient
- `\Magento\Contact\Api\MailInterface`  allows reuse of email sending in different controllers

schmengler committed:
Add docblock comments and empty lines

schmengler committed:
Add missing preference

vrann commented on app/code/Magento/Contact/Api/MailInterface.php:
```diff
@@ -0,0 +1,17 @@
+     * Send email from contact form
+     *
+     * @param string $replyTo Reply-to email address
+     * @param array $variables Email template variables
```
Service Contracts does not allow usage of "array" type, it should be more specific, i.e. "String[]"

schmengler replied:
but email template variables can be anything, from strings and integers to objects. How would you deal with it? `mixed[]` is as unspecific as `array`, so the only thing I can come up with is to hide the unspecifiy behind a `DataObject`.

schmengler replied:
I now understand that these rules only apply to service contracts in the sense of interfaces that can be exposed as web API, so I assume that I can leave the signature as it is, after moving the interfaces out of `/Api/`

vrann commented on app/code/Magento/Contact/Api/ConfigInterface.php:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Return email recipient address
+     *
+     * @return mixed
```
Service Contracts does not allow usage of "mixed" type

schmengler replied:
good catch, should have been `string` anyway

vrann commented on app/code/Magento/Contact/Api/ConfigInterface.php:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Contact module configuration
+ */
+interface ConfigInterface
```
This interface is neither Service Interface (does not describe operations on data) nor Data Interface (it does not consist only of getters and setters for the fields). Also, it does not looks like it makes sense to expose it as Web API.

Let's move these two interfaces out of Api namespace to the Model. Please add @api annotation to the interface to indicate that this is backward compatible module's interface.

schmengler replied:
Thank you for the review. I did not intend to expose the interfaces to the web api, they were only meant as service contracts. I was not aware of the convention to distinguish between `Api` namespace and `@api`  annotation, will change it as suggested

vrann reviewed (changes requested):
The new interfaces do not really belong to Api namespace because they are not service contracts and are not intended to be exposed as Web API. Please move them to the Models namespace and add @api annotation. 

There is also one more code style fix with unnecessary space.

Thanks!

PieterCappelle commented:
Some extra ideas:

1) A good next thing would be to send 2 e-mails. One to the admin and one to the submitter with a thank you message. 
1b) nl2br from comment message in email

2) Google Analytics - conversion integration

3) History of contacts in admin, possibility to export to CSV, ...

4) Enable phone field setting

*justmy2cents*

schmengler commented:
@PieterCappelle this is a bit out of scope for the refactoring. I'm trying not to add new features right now,  But that's good material for a follow-up PR!

schmengler commented on app/code/Magento/Contact/Api/ConfigInterface.php:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Return email recipient address
+     *
+     * @return mixed
```
do they allow union types like `string|string[]`?

schmengler committed:
Move contact form interfaces from Api to Model namespace

schmengler committed:
Make method signature of Mail::send() more precise

okorshenko commented:
@schmengler Thank you for your contribution and this perfect PR!
